crypto/encode_decode/decoder_pkey.c

```diff
@@ -111,6 +116,23 @@ static int decoder_construct_pkey(OSSL_DECODER_INSTANCE *decoder_inst,
     object_ref = p->data;
     object_ref_sz = p->data_size;
 
+    if(ec_flag == 1){
+        /* 
+         * When param has type "EC", we need an additional check to see if the 
+         * underlying curve is SM2. If it is SM2, we need to change the value 
+         * of data->object_type, so that we can locat the correct evp_keymgmt
+         * load function.
+         */
+        const EC_KEY* ec_object_ref = *(EC_KEY **)object_ref;
+        int curve_id = EC_GROUP_get_curve_name(EC_KEY_get0_group(ec_object_ref));
+        if(curve_id == NID_sm2){
+            OPENSSL_free(data->object_type);
+            const char* k_sm2_object_type = "SM2\0";
+            data->object_type = OPENSSL_malloc(strlen(k_sm2_object_type));
+            memcpy(data->object_type, k_sm2_object_type, strlen(k_sm2_object_type));
+        }
+    }
+
```

Comment:
Why is this done here and not in `providers/implementations/encode_decode/decode_der2key.c`

Reply:
How about around here (directly in `der2key_decode`?

https://github.com/openssl/openssl/blob/c07a34e18b098b77ce7ecb14273b7c75f59b5871/providers/implementations/encode_decode/decode_der2key.c#L315-L318

The trouble with what you're doing in `decoder_pkey.c` is that you *assume* that the decoder handles an `EC_KEY`...  but a completely different provider could have a completely different internal structure, and then this code will do a *glorious **kaboom***.  Therefore, it's better to do this hackery in actual provider code.

Reply:
The trouble with what you're doing in `decoder_pkey.c` is that you *assume* that the decoder handles an `EC_KEY`...  but a completely different provider could have a completely different internal structure, and then this code will do a *glorious **kaboom***.  Therefore, it's better to do this hackery in actual provider code.

Reply:
> How about around here (directly in `der2key_decode`)?
> 
> https://github.com/openssl/openssl/blob/c07a34e18b098b77ce7ecb14273b7c75f59b5871/providers/implementations/encode_decode/decode_der2key.c#L315-L318

hahah yep, I just also found out modifying this line would work, too.

Reply:
Yup